### PR TITLE
Fix insecure BiometricPrompt usage in Credentials class

### PR DESCRIPTION
**Pull request by bruhdev1290**
Summary
This pull request addresses the insecure usage of the BiometricPrompt API in the Mattermost Android application, identified in HackerOne report #2503621. While biometric authentication is not fully implemented in the current application, this fix prevents vulnerable code from being sent to production if biometrics are fully developed as a feature in the future. The changes ensure the secure implementation of biometric authentication by providing a valid CryptoObject parameter and improving error handling. This maintains the existing functionality of the app while enhancing security to prevent potential bypasses. For more details on the sensitivity and impact of these changes, refer to the HackerOne ticket or the manual disclosure email.

Changes
✅ implemented secure usage of the BiometricPrompt API by providing a valid CryptoObject parameter.
✅Improved error handling to prevent bypassing biometric authentication.
✅Added caching mechanism to improve performance and security.

Affected Files
com/mattermost/helpers/Credentials.java

patches/react-native-keychain+8.2.0.patch

Timeline
2024-05-14 01:08:11 UTC: Initial report (#2503621) submitted by the reporter on HackerOne outlining the insecure use of BiometricPrompt API in the Mattermost Android application.
2024-05-14 02:39:00 UTC: The reporter provides additional references and materials for understanding and reproducing the issue.
2024-05-16 08:06:49 UTC: HackerOne triage team acknowledges receipt of the report and begins review.
2024-05-16 08:42:09 UTC: @h1_analyst_jan responds, initially closing the report as informative, stating that the issue does not present a significant security risk under their current policy.
2024-05-16 12:12:11 UTC: The reporter responds, addressing the concerns raised by the triage team, emphasizing the potential risk even without physical access or rooted devices.
2024-05-16 12:14:22 UTC: The reporter clarifies the intent to prevent the vulnerable code from being fully implemented in production.
2024-05-17 21:39:09 UTC: The reporter attempts to dispute the decision with the information given above, but receives no response from the triage team nor the manual responsible disclosure email.

For additional details, refer to the original HackerOne ticket (#2503621) or the manual disclosure email.

✅ The code follows the code style of this project.
 Documentation has been updated to reflect the changes.
✅ All new and existing tests passed.
N/A Added or updated unit tests (required for all new features)
 N/A Has UI changes
 N/A Includes text changes and localization file updates
N/A Have tested against the 5 core themes to ensure consistency between them.
Device Information

This PR was tested on:
Samsung Tab A7 Lite, Android 14, and Google Pixel 7a

```release-note
Improves Handling of Authentication of the Android Client
```

### CODE REVIEW

File: android/app/src/main/java/com/mattermost/helpers/Credentials.java [hunk 1]
```diff
@@ -1,26 +1,79 @@
...
+        // keyStore.load(null);
+        // Key key = keyStore.getKey("your-key-alias", null);
+        // Cipher cipher = Cipher.getInstance("AES/CBC/PKCS7Padding");
+        // cipher.init(Cipher.ENCRYPT_MODE, key);
+        // return cipher;
+    }
+
     public static void getCredentialsForServer(ReactApplicationContext context, String serverUrl, ResolvePromise promise) {
-        final KeychainModule keychainModule = new KeychainModule(context);
+        authenticateUserWithBiometrics(context, new BiometricPrompt.AuthenticationCallback() {
```
Comment (enahum) [inline android/app/src/main/java/com/mattermost/helpers/Credentials.java:52]: This does not seem optional, also irrc some functions from push notifications do make use of this function as well as createThumbnail for videos and I'm not sure this does not disrupt the current flow.

File: patches/react-native-keychain+8.2.0.patch [hunk 2]
```diff
@@ -167,3 +167,4 @@ index 6846468..bc03596 100644
    function hasInternetCredentials(server: string): Promise<false | Result>;
  
    function setInternetCredentials(
```
Comment (enahum) [inline patches/react-native-keychain+8.2.0.patch:169]: There seems to be no changes in this file

**Review by enahum — CHANGES_REQUESTED** [review]

File: android/app/src/main/java/com/mattermost/helpers/Credentials.java [hunk 3]
```diff
@@ -1,26 +1,70 @@
...
+import java.util.concurrent.Executors;
 
 public class Credentials {
-    public static void getCredentialsForServer(ReactApplicationContext context, String serverUrl, ResolvePromise promise) {
+
+    public static void getCredentialsForServer(ReactApplicationContext context, String serverUrl, ResolvePromise promise, boolean useBiometrics) {
         final KeychainModule keychainModule = new KeychainModule(context);
 
-        final WritableMap options = Arguments.createMap();
-        // KeyChain module fails if `authenticationPrompt` is not set
```
Comment (larkox) [inline android/app/src/main/java/com/mattermost/helpers/Credentials.java:16]: Any strong reason to remove this comment?

File: android/app/src/main/java/com/mattermost/helpers/Credentials.java [hunk 4]
```diff
@@ -1,26 +1,70 @@
...
+                @Override
+                public void onAuthenticationSucceeded(BiometricPrompt.AuthenticationResult result) {
+                    WritableMap options = Arguments.createMap();
+                    WritableMap authPrompt = Arguments.createMap();
+                    authPrompt.putString("title", "Authenticate to retrieve secret");
+                    authPrompt.putString("cancel", "Cancel");
+                    options.putMap("authenticationPrompt", authPrompt);
+                    options.putString("service", serverUrl);
+
+                    keychainModule.getGenericPasswordForOptions(options, promise);
```
Comment (larkox) [inline android/app/src/main/java/com/mattermost/helpers/Credentials.java:31]: This block is the same as the block in lines 45-52. Any chance we can refactor it out?

File: patches/react-native-keychain+8.2.0.patch.o9ATfpc
Comment (larkox) [inline patches/react-native-keychain+8.2.0.patch.o9ATfpc:1]: Are these patches files intended or a bad commit?

**Review by larkox — CHANGES_REQUESTED** [review]
The idea looks good, but a few comments.

File: patches/react-native-keychain+8.2.0.patch.o9ATfpc
Comment (larkox) [reply patches/react-native-keychain+8.2.0.patch.o9ATfpc:1]: There are still some patch files that I don't understand in the changelog. Are these intended?

**Comment by bruhdev1290** [comment]
> Also, what is with all the patch files...

I needed the excess patch file due to it not compiling on my end however it was due to system configuration on my end that caused the problem I removed the excess patch files in the latest commit

**Comment by enahum** [comment]
I'm going to close this PR as is not really fixing anything nor adding functionality

**Pull request closed without merging**